service/chain/follower.go

```diff
@@ -0,0 +1,316 @@
+// Copyright 2021 Optakt Labs OÜ
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not
+// use this file except in compliance with the License. You may obtain a copy of
+// the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+package chain
+
+import (
+	"errors"
+	"fmt"
+	"math"
+
+	"github.com/dgraph-io/badger/v2"
+	"github.com/onflow/flow-go/follower"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/storage"
+	"github.com/onflow/flow-go/storage/badger/operation"
+	"github.com/rs/zerolog"
+
+	"github.com/optakt/flow-dps/models/dps"
+)
+
+type Follower struct {
+	log zerolog.Logger
+
+	consensus follower.ConsensusFollower
+	db        *badger.DB
+
+	availableHeight uint64
+	height          uint64
+	blockID         flow.Identifier
+}
+
+func FromFollower(log zerolog.Logger, follower follower.ConsensusFollower, db *badger.DB) *Follower {
+	f := Follower{
+		log: log,
+
+		consensus:       follower,
+		height:          math.MaxUint64,
+		availableHeight: math.MaxUint64,
+		blockID:         flow.ZeroID,
+	}
+
+	follower.AddOnBlockFinalizedConsumer(f.OnBlockFinalized)
+
+	return &f
+}
+
+func (f *Follower) OnBlockFinalized(finalizedBlockID flow.Identifier) {
+	var height uint64
+	err := f.db.View(operation.RetrieveFinalizedHeight(&height))
+	if err != nil {
+		f.log.Error().Err(err).Msg("Could not retrieve finalized block height")
+		return
+	}
```

Comment:
Why not just use the height on the block? Why do we need to retrieve it from the DB? When a block is signalled as finalized, all DB operations should already be done.

Reply:
WDYM the height on the block? We only get a block ID in the callback, not a block.